src/agent/ppo.py

```diff
@@ -0,0 +1,315 @@
+            
+            next_return = returns[t]
+            next_value = values[t]
+            next_advantage = advantages[t]
+        
+        # Normalize advantages
+        if len(advantages) > 1:
+            advantages = (advantages - advantages.mean()) / (advantages.std() + 1e-8)
+        
+        # Update episode data with new advantages and returns
+        for i in range(len(episode)):
+            episode[i] = PPOTransition(
+                state=episode[i].state,
+                action=episode[i].action,
+                log_prob=episode[i].log_prob,
+                value=episode[i].value,
+                reward=episode[i].reward,
+                done=episode[i].done,
+                next_state=episode[i].next_state,
+                next_acts=episode[i].next_acts,
+                advantage=advantages[i],
+                returns=returns[i]
+            )
+        
+        return advantages, returns
+
+    def update(self):
+        """Update policy using PPO"""
+        if not self.completed_episodes:
+            return 0.0  # No data to update from
+        
+        total_loss = 0
+        
+        # Now process all completed episodes
+        for episode in self.completed_episodes:
+            # Compute advantages for this episode if not already done
+            if episode[0].advantage is None:
+                self._compute_truncated_advantages(episode, episode[-1].done)
+            
+            # Get episode data
+            states = [t.state for t in episode]
+            actions = torch.tensor([t.action for t in episode], device=device, dtype=torch.long)
+            old_log_probs = torch.tensor([t.log_prob for t in episode], device=device, dtype=torch.float32)
+            advantages = torch.tensor([t.advantage for t in episode], device=device, dtype=torch.float32)
+            returns = torch.tensor([t.returns for t in episode], device=device, dtype=torch.float32)
+            
+            # Multiple optimization epochs
+            for _ in range(self.ppo_epochs):
+                state_batch = states
+                action_scores, values = self.network(state_batch)
+                dist = Categorical(F.softmax(action_scores, dim=-1))
+                new_log_probs = dist.log_prob(actions)
+                entropy = dist.entropy().mean()
+                
+                # Calculate ratios and surrogate losses
+                ratios = torch.exp(new_log_probs - old_log_probs)
+                surr1 = ratios * advantages
+                surr2 = torch.clamp(ratios, 1-self.clip_epsilon, 1+self.clip_epsilon) * advantages
+                
+                # PPO policy and value losses
+                policy_loss = -torch.min(surr1, surr2).mean()
+                value_loss = F.mse_loss(values.squeeze(), returns)
+                entropy_loss = -self.entropy_coef * entropy
+                loss = policy_loss + 0.5 * value_loss + entropy_loss
```

Comment:
PPO Lagrangian should be pretty straightforward here... its just adding an optimizer for the $$\lambda$$ parameter and the cost term here.